### PR TITLE
#167730727 Setup Postgres 

### PR DESCRIPTION
**Pull request by dewaleolaoye**
#### What does this PR do?
Configure BareFoot Nomad to use PostgreSQL as DB Engine
#### Description of Task to be completed?
* Setup Sequelize
#### How should this be manually tested?
* git clone this branch [ch-setup-postgres-167730728](https://github.com/andela/firestar-backend/tree/ch-setup-postgres-167730727)
* cd to the project
* Checkout to the branch
* Check `package.json` file to see `pg` dependency

#### Any background context you want to provide?
* N/A

#### What are the relevant pivotal tracker stories?
[#167730727](https://www.pivotaltracker.com/story/show/167730727)

#### Screenshots or gifs (if appropriate)
* N/A
#### Questions:
* N/A

### CODE REVIEW

**Review by conquext — CHANGES_REQUESTED** [review]
This file should go in src/database/config

Also separate configuration concerns from migration services.

**Review by marusoft — APPROVED** [review]
awesome work, work as expected

File: src/database/config.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+const dotenv = require('dotenv');
+dotenv.config();
+
+const pool = new Pool({
+  connectionString: process.env.DATABASE_URL,
+});
+
+pool
+  .connect()
+  .then(() => console.log('connected to db'))
```
[the remark this answers is not in the document]
Comment (conquext) [reply src/database/config.js]: /* Great job centralizing all access to the database to one file
 * You now have to expose the global pool to the app
 */
module.exports = {
   query: function(text, values, cb) {
      pool.connect(function(err, client, done) {
        client.query(text, values, function(err, result) {
          done();
          cb(err, result);
        })
      });
   }
}

This way you can be able to
1. instantiate the pool in one spot in the app and expose it to all queries.
2. handle 1 query at a time per connection. 
3. control the life-cycle of all connections.
4. open up a 1-off client to kill some hung stuff.
5. share a connection across multiple queries within a request and only have to change things in one place.

You might want to change your call to pool.connect() now to listen for connections instead
i.e. 
pool.on('connect', () => {
  debug('connected to the Database');
});

Also, create a separate database url for test or development environment.

**Review by conquext — CHANGES_REQUESTED** [review]
/* Great job centralizing all access to the database to one file
 * You now have to expose the global pool to the app by exporting your setup
 */
module.exports = {
   query: function(text, values, cb) {
      pool.connect(function(err, client, done) {
        client.query(text, values, function(err, result) {
          done();
          cb(err, result);
        })
      });
   }
}

This way you can be able to
1. instantiate the pool in one spot in the app and expose it to all queries.
2. handle 1 query at a time per connection. 
3. control the life-cycle of all connections.
4. open up a 1-off client to kill some hung stuff.
5. share a connection across multiple queries within a request and only have to change things in one place.

You might want to change your call to pool.connect() now to listen for connections instead
i.e. 
pool.on('connect', () => {
  debug('connected to the Database');
});

Also, create a separate database url for test or development environment.

**Comment by dewaleolaoye** [comment]
@conquext Thanks, I will implement it now

**Comment by conquext** [comment]
@dewaleolaoye This is fine, but what if we need to check out a client from the pool to run several queries in a row in a transaction?

Since node only loads the module once when it's required so only one client object exists and is shared by your entire app. As you're creating the connection manually once at startup then if it ever has an issue (say the server reboots or the socket has a network issue and closes) the app may never recover. 
So I will suggest we tie our connection to the pool to the lifecycle of their use. If it's tied to a request, or the part of the code that's performing an action, then create the client in that function and dispose of it when that piece is done. The psuedocode would be:

`module.exports = function query(text, params) {
  return new Promise((resolve, reject) => {
    try {
      const client = pool.connect();
      client.query(text, params)
        .then((res) => {
          resolve(res);
        }).catch((error) => {
          reject(error);
        })
        .finally(() => { client.release() ;});
    } catch (e) {
      reject(e);
    }
  });
};
`

File: src/database/config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dewaleolaoye) [reply src/database/config.js]: @conquext  Okay! but in this case, what we need here is to install ***pg*** module and establish a connection with the database. Sequelize handles every other thing needed like query.

**Review by conquext — APPROVED** [review]
I got clarifications that the task requirement was to setup postgres as a db engine for sequelize ORM. So it is clear we won't need to separately configure postgres. This PR is ok to merge.

File: src/database/config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (rovilay) [reply src/database/config.js]: @dewaleolaoye Please resolve outdated conversations

[the remark this answers is not in the document]
Comment (dewaleolaoye) [reply src/database/config.js]: @rovilay Okay, I will fix it now

**Pull request merged**